### PR TITLE
CT-637 - Add EOS recovery support

### PR DESCRIPTION
**Pull request by CapnMigraine**
https://bitgoinc.atlassian.net/browse/CT-637

### CODE REVIEW

File: modules/core/src/v2/coins/eos.ts [hunk 1]
```diff
@@ -91,6 +93,29 @@ interface ExplainTransactionOptions {
   headers: EosTransactionHeaders;
 }
 
+interface SendableTransaction {
+  compression: string;
+  packed_trx: string;
+  signatures: string[];
+}
```
Comment (marktoda) [inline modules/core/src/v2/coins/eos.ts]: \n

File: modules/core/src/v2/coins/eos.ts [hunk 2]
```diff
@@ -91,6 +93,29 @@ interface ExplainTransactionOptions {
...
+interface RecoveryTransaction extends SendableTransaction {
```
Comment (marktoda) [inline modules/core/src/v2/coins/eos.ts]: Why extend sendabletransaction but also have a sendable transaction inside? Maybe RecoveryTransaction should be a basecoin class, extended here as EosRecoveryTransaction

File: modules/core/src/v2/coins/eos.ts [hunk 3]
```diff
@@ -124,6 +149,12 @@ export class Eos extends BaseCoin {
     return true;
   }
 
+  /**
+   * Get URL of a public node
+   */
+  getPublicNodeUrl(): string {
+    return 'https://bp.cryptolions.io';
```
Comment (marktoda) [inline modules/core/src/v2/coins/eos.ts]: maybe a few different options for each network in case one is down? If that's too annoying we can do it later if it becomes a problem

File: modules/core/src/v2/coins/eos.ts [hunk 4]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
       .call(this)
       .asCallback(callback);
   }
+
+  /**
+   * Validate a public or private key
+   * If passphrase is provided, try to decrypt the key with it
+   * @param key
```
Comment (marktoda) [inline modules/core/src/v2/coins/eos.ts]: types?

File: modules/core/src/v2/coins/eos.ts [hunk 5]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
...
+   * @param source
+   * @param passphrase
+   * @param isUnsignedSweep
+   * @param isKrsRecovery
+   */
+  validateKey = function({ key, source, passphrase, isUnsignedSweep, isKrsRecovery }: ValidateKeyParams): HDNode {
```
Comment (marktoda) [inline modules/core/src/v2/coins/eos.ts]: types for params

File: modules/core/src/v2/coins/eos.ts [hunk 6]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
...
+  getTransactionHeadersFromNode(): Bluebird<any> {
+    return co(function*() {
+      const nodeUrl = this.getPublicNodeUrl();
+      const chainInfoResult = yield this.bitgo.post(nodeUrl + '/v1/chain/get_info').send();
+      if (!chainInfoResult || chainInfoResult.status !== 200) {
+        throw new Error('unable to fetch chain info from node: ' + nodeUrl);
+      }
+      const chainInfo = chainInfoResult.body;
+
+      const headBlockInfoResult = yield this.bitgo.post(nodeUrl + '/v1/chain/get_block').send({
```
Comment (marktoda) [inline modules/core/src/v2/coins/eos.ts]: Since you're already using a public node (i.e. assuming that bitgo is down), can we just use the get_block EOS api to get this info?

File: modules/core/src/v2/coins/eos.ts [hunk 7]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
...
+
+      return {
+        expiration: expirationDate.toISOString(),
+        ref_block_num: chainInfo.head_block_num & 0xffff,
+        ref_block_prefix: headBlockInfoResult.body.ref_block_prefix,
+      };
+    }).call(this);
+  }
+
+  getTransferAction = function({ recipient, sender, amount, memo }) {
```
Comment (marktoda) [inline modules/core/src/v2/coins/eos.ts]: Seeing the amount of stuff that is copied in platform and here makes me wonder how we can refactor for reuse -- it'd be nice if maybe Statics could have utility functions like this for each coin. Just a thought, not relevant to this PR

File: modules/core/src/v2/coins/eos.ts [hunk 8]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
...
+        to: recipient,
+        quantity: `${this.baseUnitsToBigUnits(amount)} EOS`,
+        memo: !_.isNil(memo) ? memo : '', // Memo must be defined, set it to empty string if it is not
+      },
+    };
+  };
+
+  /**
+   * Sign a transaction with a key
+   * @param tx
```
Comment (marktoda) [inline modules/core/src/v2/coins/eos.ts]: There are so many different types of `tx` objects in EOS I think we should be as specific as possible -- like "signableTxString" or something lol. "tx" could be the packed_trx, or this, or the stringified send JSON -- just too confusing

File: modules/core/src/v2/coins/eos.ts (hunk 4 above)
Comment (tylerlevine) [reply modules/core/src/v2/coins/eos.ts]: Typedoc will take them from the typescript types, so no need.

File: modules/core/src/v2/coins/eos.ts (hunk 5 above)
Comment (tylerlevine) [reply modules/core/src/v2/coins/eos.ts]: Types are in `ValidateKeyParams`. However, the pattern we have  been following is to use the suffix `Options`,  so this should be `ValidateKeyOptions`.

File: modules/core/src/v2/coins/eos.ts [hunk 9]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
...
+      ];
+      if (!this.isValidAddress(recoveryDestination)) {
+        throw new Error('Invalid destination address!');
+      }
+
+      return keys;
+    }).call(this);
+  }
+
+  getAccountFromNode(address): Bluebird<any> {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: Doc comment please, also add types

File: modules/core/src/v2/coins/eos.ts [hunk 10]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
...
+      const nodeUrl = this.getPublicNodeUrl();
+      const result = yield this.bitgo.post(nodeUrl + '/v1/chain/get_account').send({ account_name: address });
+      if (!result || result.status !== 200) {
+        throw new Error('unable to fetch account from node: ' + nodeUrl);
+      }
+      return result.body;
+    }).call(this);
+  }
+
+  getTransactionHeadersFromNode(): Bluebird<any> {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: Doc comment

File: modules/core/src/v2/coins/eos.ts (hunk 7 above)
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: No need for `= function`

File: modules/core/src/v2/coins/eos.ts [hunk 11]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
...
+      },
+    };
+  };
+
+  /**
+   * Sign a transaction with a key
+   * @param tx
+   * @param signingKey
+   */
+  signTx = function(tx: string, signingKey: HDNode) {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: No need for `= function`, please add return type

File: modules/core/src/v2/coins/eos.ts [hunk 12]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
...
+    return ecc.Signature.sign(signBuffer, privateKeyBuffer).toString();
+  };
+
+  /**
+   * Serialize an EOS transaction, to the format that should be signed
+   * @param eosClient an offline EOSClient that has the transaction structs
+   * @param transaction The EOS transaction returned from `eosClient.transaction` to serialize
+   * @return {String} serialized transaction in hex format
+   */
+  serializeTransaction = function(eosClient, transaction) {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: No need for `= function`, please add types

File: modules/core/src/v2/coins/eos.ts [hunk 13]
```diff
@@ -408,4 +438,268 @@ export class Eos extends BaseCoin {
...
+
+    return EosJs.modules.Fcbuffer.toBuffer(eosTxStruct, txObject).toString('hex');
+  };
+
+  /**
+   * Builds a funds recovery transaction without BitGo
+   * @param params
+   * @param callback
+   */
+  recover(params: RecoveryOptions, callback): Bluebird<RecoveryTransaction> {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: callback param should be `callback?: NodeCallback<RecoveryTransaction>`

File: modules/core/src/v2/coins/teos.ts [hunk 14]
```diff
@@ -18,4 +18,8 @@ export class Teos extends Eos {
   getFullName() {
     return 'Testnet EOS';
   }
+
+  getPublicNodeUrl(): string {
+    return 'https://jungle2.cryptolions.io';
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/teos.ts]: consider adding to environments.ts instead of adding to the coin class.

File: modules/core/src/v2/coins/eos.ts [hunk 15]
```diff
@@ -408,4 +441,315 @@ export class Eos extends BaseCoin {
...
+   * @param passphrase
+   * @param isUnsignedSweep
+   * @param isKrsRecovery
+   */
+  validateKey({ key, source, passphrase, isUnsignedSweep, isKrsRecovery }: ValidateKeyOptions): HDNode {
+    if (!key.startsWith('xprv') && !isUnsignedSweep) {
+      // Try to decrypt the key
+      try {
+        if (source === 'user') {
+          return HDNode.fromBase58(sjcl.decrypt(passphrase, key));
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: Please use `this.bitgo.decrypt()` instead of calling sjcl directly

File: modules/core/src/v2/coins/eos.ts [hunk 16]
```diff
@@ -408,4 +441,315 @@ export class Eos extends BaseCoin {
...
+      return keys;
+    }).call(this);
+  }
+
+  /**
+   * Make a request to one of the public EOS nodes available
+   * @param endpoint
+   * @param payload
+   */
+  getDataFromNode({ endpoint, payload }: { endpoint: string; payload?: object }): Bluebird<any> {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: Can this be made protected? Or does this need to be part of the public API?

Same for other helper methods below - I'd rather not expose these unless really needed.

File: modules/core/src/v2/coins/xlm.ts [hunk 17]
```diff
@@ -46,15 +42,7 @@ interface VerifyAddressOptions {
...
-interface InitiateRecoveryOptions {
-  userKey: string;
-  backupKey: string;
-  recoveryDestination: string;
-  krsProvider?: string;
-  walletPassphrase?: string;
-}
-
-interface RecoveryOptions extends InitiateRecoveryOptions {
+interface RecoveryOptions extends BaseCoinInitiateRecoveryOptions {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/xlm.ts]: Let's do this in a separate PR please.

File: modules/core/src/v2/coins/eos.ts [hunk 18]
```diff
@@ -1,19 +1,22 @@
...
 import * as url from 'url';
 import * as querystring from 'querystring';
 import * as _ from 'lodash';
-import { InvalidAddressError, UnexpectedAddressError } from '../../errors';
 import * as Bluebird from 'bluebird';
 const co = Bluebird.coroutine;
+import { InvalidAddressError, UnexpectedAddressError } from '../../errors';
+import * as config from '../../config';
+import * as sjcl from '../../vendor/sjcl.min.js';
+import * as common from '../../common';
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: Please import `Environments` from `v2/environments.ts` instead of importing through common.

File: modules/core/src/v2/coins/eos.ts [hunk 19]
```diff
@@ -124,6 +152,12 @@ export class Eos extends BaseCoin {
     return true;
   }
 
+  /**
+   * Get URLs of some active public nodes
+   */
+  getPublicNodeUrls(): string[] {
+    return common.Environments[this.bitgo.env].eosNodeUrls;
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: Please use `this.bitgo.getEnv()` instead of accessing a private property of the bitgo object.

File: modules/core/src/v2/coins/eos.ts [hunk 20]
```diff
@@ -1,19 +1,22 @@
...
 import * as ecc from 'eosjs-ecc';
 import * as url from 'url';
 import * as querystring from 'querystring';
 import * as _ from 'lodash';
-import { InvalidAddressError, UnexpectedAddressError } from '../../errors';
 import * as Bluebird from 'bluebird';
 const co = Bluebird.coroutine;
+import { InvalidAddressError, UnexpectedAddressError } from '../../errors';
+import * as config from '../../config';
+import * as sjcl from '../../vendor/sjcl.min.js';
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: Remove and use `bitgo.encrypt()` or `bitgo.decrypt()` instead.

File: modules/core/src/v2/coins/eos.ts (hunk 6 above)
Comment (marktoda) [reply modules/core/src/v2/coins/eos.ts]: oh man yep my bad.. thought this was hitting the ims
